### PR TITLE
more generic broadcast over a single sparse matrix

### PR DESCRIPTION
**Pull request by Sacha0**
In #18590, one proposal for a more generic `broadcast` over sparse matrices received broad support: At runtime, (1) determine whether the `broadcast` operation returns zero when all (relevant) arguments are zero (by sampling `f(0...)`), (2) branch to code specialized for the case at hand, and (3) in either case return a sparse matrix for type stability.

This pull request begins implementing that proposal. Specifically, this pull request starts with `broadcast` over a single sparse matrix (and also handles a few common combinations of a single sparse matrix with `broadcast` scalars).

A few code (and corresponding PR) comments highlight some decisions. The primary decision concerns #18309: When the `broadcast` operation maps a (stored) nonzero entry to zero, should the result store that zero (nz2nz behavior) or not (nz2z behavior)? I now lean towards nz2z behavior for the following reason:

Consider `C = A .* B` (and similar operations) where `A` and `B` are sparse. Chances are you want `C` to the store the intersection (rather than union) of the structures of `A` and `B` (the present behavior of `.*` and friends). Returning the structural intersection rather than union in such cases requires the nz2z behavior when `broadcast`ing over two or more sparse matrices (assuming you determine the `broadcast` operation's zero-preserving behavior solely via sampling, see #18590). Having `broadcast` over a single sparse matrix behave differently seems like a startling inconsistency. That inconsistency seems more important than the performance and implementation complexity differences mentioned in #18309. Regarding the zero-purging consistency point in #18309, with more thought this part of `broadcast`'s behavior isn't really zero-purging in the sense discussed elsewhere (not so sure about `broadcast!` though).

Thoughts? Best!

### CODE REVIEW

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -1390,7 +1390,77 @@ function one{T}(S::SparseMatrixCSC{T})
...
+    Bnzval = broadcast(f, A.nzval); resize!(Bnzval, nnz(A))
+    return SparseMatrixCSC(A.m, A.n, Bcolptr, Browval, Bnzval)
+    # TODO: The following lines for constructing Bcolptr and Browval might be safer than
+    # those above: If A.colptr and/or A.rowval are shorter than their expected lengths, the
+    # copy!s might catch the error, whereas the former copy/resize! combinations won't. The
+    # lines above are simpler/cleaner though. Thoughts?
+    # Bcolptr = similar(A.colptr, A.n + 1)
+    # copy!(Bcolptr, 1, A.colptr, 1, A.n + 1)
+    # Browval = similar(A.rowval, nnz(A))
+    # copy!(Browval, 1, A.rowval, 1, nnz(A))
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: These alternative lines for constructing `Bcolptr` and `Browval` might be safer than the `copy`/ `resize!` lines above: If `A.colptr` and/or `A.rowval` are shorter than their expected lengths, the `copy!`s might catch the error, whereas the former `copy`/`resize!` combinations won't. The `copy`/`resize!` lines are simpler/cleaner though. Thoughts? Thanks!

These alternative lines for constructing `Bcolptr` and `Browval` might be safer

File: base/sparse/sparsematrix.jl [hunk 2]
```diff
@@ -1390,7 +1390,77 @@ function one{T}(S::SparseMatrixCSC{T})
...
+# (2) any combination (number, order, type mixture) of broadcast scalars. One might achieve
+# (2) via a generated function. But how one might simultaneously and gracefully achieve (1)
+# eludes me. Thoughts much appreciated.
+#
+broadcast{Tf}(f::Tf, x::Union{Number,Bool}, A::SparseMatrixCSC) = broadcast(y -> f(x, y), A)
+broadcast{Tf}(f::Tf, A::SparseMatrixCSC, y::Union{Number,Bool}) = broadcast(x -> f(x, y), A)
+# NOTE: The following two method definitions work around #19096. These definitions should
+# be folded into the two preceding definitions on resolution of #19096.
+broadcast{Tf,T}(f::Tf, ::Type{T}, A::SparseMatrixCSC) = broadcast(y -> f(T, y), A)
+broadcast{Tf,T}(f::Tf, A::SparseMatrixCSC, ::Type{T}) = broadcast(x -> f(x, T), A)
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: The four method definitions above for handling common combinations of sparse matrices with `broadcast` scalars is not satisfying: A better solution would achieve the same for (1) all `broadcast` scalar types (`Base.Broadcast.containertype(x) == Any`?) and (2) any combination (number, order, type mixture) of `broadcast` scalars. One might achieve (2) via a generated function. But how one might simultaneously and gracefully achieve (1) eludes me. Thoughts much appreciated. Thanks!

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -1390,7 +1390,77 @@ function one{T}(S::SparseMatrixCSC{T})
...
+
+# TODO: Determine what to do about collapsing the nz2nz_z2z and nz2z_z2z classes. Either (1)
+# the generic unary sparse broadcast behavior should be nz2z_z2z instead of the nz2z_z2z
+# behavior presently above (this would be consistent with existing binary sparse broadcast
+# behavior); or (2) the generic unary broadcast behavior should be nz2nz_z2z as above, and the
+# nz2z_z2z specializations below should ultimately go away in favor of `broadcast`+`dropzeros!`
+# (this would not be consistent with existing binary sparse broadcast behavior).
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: This TODO concerns the primary decision described in the original post.

File: base/sparse/sparsematrix.jl [hunk 4]
```diff
@@ -1390,7 +1390,77 @@ function one{T}(S::SparseMatrixCSC{T})
...
 
 """
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Once the nz2z/nz2nz decision is made, most of the code below this line till `## Broadcasting kernels specialized for returning a SparseMatrixCSC` should go away.

**Comment by stevengj** [comment]
My initial preference would be `nz2z` (don't store zeros) for `broadcast`, but `nz2nz` (store zeros) for `broadcast!`, if that is feasible.

**Comment by stevengj** [comment]
On the other hand, implementation would be simpler and faster if we used `nz2nz` everywhere for unary operations, and in practice I'm very skeptical that any real user will have a problem with this.

**Comment by Sacha0** [comment]
> My initial preference would be nz2z (don't store zeros) for broadcast, but nz2nz (store zeros) for broadcast!, if that is feasible.

Cheers, my thinking as well. I appreciate your devil's-advocate arguments though; thoughts below.

> On the other hand, implementation would be simpler and faster if we used nz2nz everywhere for unary operations

`nz2nz`'s implementation simplicity is appealing. On the other hand, the code complexity / length differences between `nz2nz`'s implementation and a simplified `nz2z` implementation seem negligible relative to consistency and ideal-behavior concerns.

The benchmarks in #18309 indicate that `nz2nz` is somewhat faster in some (common?) cases (typically by a few percent, maximally by about thirty percent). On the other hand, `nz2z` is comparably faster in other (common?) cases, and *much* faster in still other (uncommon?) cases. In other words, `nz2z`'s performance is mostly comparable to that of `nz2nz`, but `nz2z`'s worst-case performance is meaningfully better than `nz2nz`'s, perhaps making it a good choice for a generic method.

Another small argument for `nz2z`: If a developer needs `nz2nz` behavior, she should be able to safely achieve it in less than five lines of code, whereas `nz2nz` is more subtle.

Best!

**Comment by Sacha0** [comment]
(Absent additional input in the next couple of days I will revise this PR with `nz2z` behavior.)

File: base/sparse/sparsematrix.jl [hunk 5]
```diff
@@ -1390,7 +1390,77 @@ function one{T}(S::SparseMatrixCSC{T})
...
 end
 
-## Unary arithmetic and boolean operators
+
+## Broadcast operations involving a single sparse matrix and possibly broadcast scalars
+
+function broadcast{Tf}(f::Tf, A::SparseMatrixCSC)
+    fofzero = f(zero(eltype(A)))
+    fpreszero = fofzero == zero(fofzero)
+    return fpreszero ? _broadcast_zeropres(f, A) : _broadcast_notzeropres(f, A)
```
Comment (stevengj) [inline base/sparse/sparsematrix.jl]: Shouldn't you pass `fofzero` to `_broadcast_notzeropres` so that it doesn't need to be recomputed?

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Originally I passed `fofzero` to `_broadcast_notzeropres` for that reason, but later nixed the pass for simplicity. Happy to restore the pass if that strikes you as better though? Thanks!

Comment (stevengj) [inline base/sparse/sparsematrix.jl]: It shouldn't make much difference, I guess, so whatever you prefer.

**Comment by Sacha0** [comment]
Switched to `nz2z` behavior and nixed associated specialized methods. Best!

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (stevengj) [inline base/sparse/sparsematrix.jl]: I don't think you need a generated function.   Maybe we can brainstorm with @pabloferz about this at some point soon?  Maybe start a discourse thread?  @pabloferz, are you on discourse.julialang.org yet?

In any case, it seems like that should be separate from this PR, which is only about the single-argument case.

**Comment by stevengj** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19239#issuecomment-260098121) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/edc39f302685e2d7b7c049f655452296d7c4133c/8601d22_vs_42df52a/report.md). cc @jrevels

**Comment by stevengj** [comment]
Big improvement in the sparse-broadcast benchmark, unexpectedly.  Odd regression in the array-[comprehension collect](https://github.com/JuliaCI/BaseBenchmarks.jl/blob/2f3de40215cf329e917971d1f07961ad8c7e9ac9/src/array/ArrayBenchmarks.jl#L179) benchmark, probably unrelated noise since that doesn't involve sparse matrices at all.

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: Al least for this case you might need another trait layer. E.g.

```julia
import Base.Broadcast.containertype
broadcast(f, A, B::SparseMatrixCSC) = broadcast_sp(f, containertype(A), containertype(B), A, B)
broadcast(f, A::SparseMatrixCSC, B) = broadcast_sp(f, containertype(A), containertype(B), A, B)
broadcast_sp(f, ::Type{Any}, ::Type{Array} A, B) = broadcast(x -> f(A, x), B)
broadcast_sp(f, ::Type{Array}, ::Type{Any} A, B) = broadcast(x -> f(x, B), A)
broadcast_sp(f, ::Type, ::Type, A, B) = Base.Broadcast.broadcast_c(f, Array, A, B)
```

@stevengj I'm already on discourse, so feel free to ping me (`@pabloferz`) if you'd like to start a thread overe there.

**Comment by stevengj** [comment]
Seeing the same `testset spawn` error on Linux as in #19305, so seems to be unrelated.

**Comment by Sacha0** [comment]
Pushed two small modifications: (1) `broadcast` now passes `f(zero(eltype(A)))` to `_broadcast_notzeropres` via a `fillvalue` argument. (2) removed a question that should not persist in the comments. Best!

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: The following snippet works for arbitrary combinations of broadcast scalars with a single sparse matrix, but the `(y...) = f(a, y...)` kills inference.
```julia
import Base.Broadcast: broadcast_c, containertype, promote_containertype

containertype{T<:SparseMatrixCSC}(::Type{T}) = SparseMatrixCSC
# Combinations of sparse arrays with broadcast scalars should yield sparse arrays
promote_containertype(::Type{Any}, ::Type{SparseMatrixCSC}) = SparseMatrixCSC
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Any}) = SparseMatrixCSC
# Combinations of sparse arrays with anything else should fall back to dense array methods
promote_containertype(::Type{Array}, ::Type{SparseMatrixCSC}) = Array
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Array}) = Array
promote_containertype(::Type{Tuple}, ::Type{SparseMatrixCSC}) = Array
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Tuple}) = Array

broadcast_c(f, SparseMatrixCSC, A) = broadcast(f, A)
broadcast_c(f, SparseMatrixCSC, A, Bs...) = broadcast_sp(f, containertype(A), containertype(Bs...), A, Bs...)
broadcast_sp(f, ::Type{Any}, ::Type{SparseMatrixCSC}, a, Bs...) = broadcast_c((y...) -> f(a, y...), SparseMatrixCSC, Bs...)
broadcast_sp(f, ::Type{SparseMatrixCSC}, ::Type{Any}, A, bs...) = broadcast_c(x -> f(x, bs...), SparseMatrixCSC, A)
broadcast_sp(f, ::Type{SparseMatrixCSC}, ::Type{SparseMatrixCSC}, A, Bs...) = broadcast_c(f, Array, A, Bs...)
```

Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: What about

```julia
broadcast_sp(f, ::Type{Any}, ::Type{SparseMatrixCSC}, a, Bs...) = broadcast_c(y -> f(a, y...), SparseMatrixCSC, Bs)
```

?

Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: A possible solution would be to have (additionally to what you have above)

```julia
immutable FArgs{F,T}
    f::F
    args::T
end
(f::FArgs)(x...) = f.f(f.args..., x...)

broadcast_sp(f, ::Type{Any}, ::Type{SparseMatrixCSC}, a, Bs...) = broadcast_c(FArgs(f,a), SparseMatrixCSC, Bs...)
broadcast_sp(f::FArgs, ::Type{Any}, ::Type{SparseMatrixCSC}, a, Bs...) = broadcast_c(FArgs(f.f,(f.args...,a)), SparseMatrixCSC, Bs...)
broadcast_sp(f::FArgs, ::Type{SparseMatrixCSC}, ::Type{Any}, A, bs...) = broadcast_c(x -> f.f(f.args..., x, bs...), SparseMatrixCSC, A)
```

Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: The hacky snippet below takes a different tack. In principle it should handle any combination of broadcast scalars and `SparseMatrixCSC`s; in practice type inference fails when the number of arguments exceeds four or more than one `SparseMatrixCSC` is involved.
```julia
import Base: tail, front
import Base.Broadcast: broadcast_c, containertype, promote_containertype, broadcast_indices

broadcast_indices(::Type{SparseMatrixCSC}, A) = indices(A)
containertype{T<:SparseMatrixCSC}(::Type{T}) = SparseMatrixCSC
# Combinations of sparse arrays with broadcast scalars should yield sparse arrays
promote_containertype(::Type{Any}, ::Type{SparseMatrixCSC}) = SparseMatrixCSC
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Any}) = SparseMatrixCSC
# Combinations of sparse arrays with anything else should fall back to dense array methods
promote_containertype(::Type{Array}, ::Type{SparseMatrixCSC}) = Array
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Array}) = Array
promote_containertype(::Type{Tuple}, ::Type{SparseMatrixCSC}) = Array
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Tuple}) = Array

broadcast_c(f, ::Type{SparseMatrixCSC}, As...) = foo((x, y) -> f(y...), (), first(As), tail(As))

foo(f, arrays, nextarg, remargs) =
    foo((x, y) -> f(x, (nextarg, y...)), arrays, first(remargs), tail(remargs))
foo(f, arrays, nextarg::SparseMatrixCSC, remargs) =
    foo((x, y) -> f(front(x), (last(x), y...)), (arrays..., nextarg), first(remargs), tail(remargs))

foo(f, arrays, nextarg, remargs::Tuple{}) =
    bar(x -> f(x, (nextarg,)), arrays...)
foo(f, arrays, nextarg::SparseMatrixCSC, remargs::Tuple{}) =
    bar(x -> f(front(x), (last(x),)), arrays..., nextarg)

bar(f, A) = broadcast(x -> f((x,)), A)
bar(f, A, B) = broadcast_c((x,y) -> f((x,y)), Array, A, B)
bar(f, A, B, Cs...) = broadcast_c((x,y,zs...) -> f((x,y,zs...)), Array, A, B, Cs...)
```
I'm sure a better approach exists, possibly e.g. inverting the call chain. Having a look at `@pabloferz`'s suggestion now. Best!

**Comment by stevengj** [comment]
The Travis problem should now be fixed, if you want to rebase.

**Comment by Sacha0** [comment]
Rebased. I've made some headway re. https://github.com/JuliaLang/julia/pull/19239#pullrequestreview-7339126, but improvements on that front could be an independent PR (however you prefer). Thanks and best!

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: New versions capable (in principle) of handling any number of `SparseMatrixCSC`s follow below. None circumvent #19096. **TL;DR** The first version (A) is the best so far, perhaps good enough to incorporate for now? Thoughts? Any and all pointers much appreciated, particularly towards better understanding of why type inference works/fails where it does.

Code common to all versions:
```julia
import Base: front, tail
import Base.Broadcast: broadcast_c, containertype, promote_containertype, broadcast_indices

# Broadcast container type promotion for SparseMatrixCSCs
containertype{T<:SparseMatrixCSC}(::Type{T}) = SparseMatrixCSC
broadcast_indices(::Type{SparseMatrixCSC}, A) = indices(A)
# Combinations of sparse arrays with broadcast scalars should yield sparse arrays
promote_containertype(::Type{Any}, ::Type{SparseMatrixCSC}) = SparseMatrixCSC
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Any}) = SparseMatrixCSC
# Combinations of sparse arrays with anything else should fall back to dense array
promote_containertype(::Type{Array}, ::Type{SparseMatrixCSC}) = Array
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Array}) = Array
promote_containertype(::Type{Tuple}, ::Type{SparseMatrixCSC}) = Array
promote_containertype(::Type{SparseMatrixCSC}, ::Type{Tuple}) = Array
```

(A) For the following version, type inference remains happy when the argument list (1) contains a single sparse matrix and up to six broadcast scalars, (2) contains two sparse matrices and up to twelve broadcast scalars (more depending on position, strangely), or (3) up to six sparse matrices as long as no broadcast scalars appear. This behavior is a mystery to me, insight much appreciated. This version seems the best in practice at the moment.
```julia
# closeoverscalars steps through the argument list. For each scalar argument,
# closeoverscalars generates a closure incorporating that scalar into the present version
# of the broadcast function, removes that scalar from the argument list, and steps to the
# next argument. For each matrix argument, closeoverscalars generates a closure which
# will later funnel that matrix argument to the right slot in the original broadcast
# function, collects that matrix in a tuple of matrices over which to eventually broadcast,
# removes that matrix from the argument list, and steps to the next argument. After
# traversing the argument list, closeoverscalars returns a closure incorporating
# all broadcast scalars in the argument list followed by the argument list's sparse matrices.
#
# Entrypoint to closeoverscalars
broadcast_c(f, ::Type{SparseMatrixCSC}, args...) = broadcast(closeoverscalars(f, args)...)
closeoverscalars(f, args) = closeoverscalars((x, y) -> f(y...), (), args...)
# Recursion cases for closeoverscalars
closescalar(f, s) = (x, y) -> f(x, (s, y...))
passmatrix(f) = (x, y) -> f(front(x), (last(x), y...))
closeoverscalars(f, mats, nextarg, remargs...) = closeoverscalars(closescalar(f, nextarg), mats, remargs...)
closeoverscalars(f, mats, nextarg::SparseMatrixCSC, remargs...) = closeoverscalars(passmatrix(f), (mats..., nextarg), remargs...)
# Base cases for closescalars
closelastscalar(f, s) = (x...) -> f(x, (s,))
passlastmatrix(f) = (x...) -> f(front(x), (last(x),))
closeoverscalars(f, mats, lastarg) = (closelastscalar(f, lastarg), mats...)
closeoverscalars(f, mats, lastarg::SparseMatrixCSC) = (passlastmatrix(f), mats..., lastarg)
```

(B) For the following version (a descendant of @pabloferz's suggestion), type inference remains happy when the argument list (1) contains up to nine members including either one or two sparse matrices, or (2) contains three sparse matrices and no broadcast scalars. This behavior is also a mystery to me, insight much appreciated.
```julia
# broadcast_sp steps through the argument list, collecting each sequence of broadcast scalars
# between one sparse matrix (or start or end of the argument list) and the next in a tuple,
# storing the sequence of such tuples in stups ("scalar-tuples"), and storing the sparse
# matrices in mats. After traversing the argument list, broadcast_sp calls closeoverscalars
# which, given stups, mats, and the broadcast function, generates a closure incorporating
# all broadcast scalars in the argument list. broadcast_sp then passes that closure along
# with the argument list's sparse matrices on to broadcast.
#
# Entrypoint to broadcast_sp
broadcast_c(fn, ::Type{SparseMatrixCSC}, A, Bs...) =
    broadcast_sp(fn, ((),), (), containertype(A), containertype(Bs...), A, Bs...)
# Recursion cases for broadcast_sp
broadcast_sp(fn, stups, mats, ::Type{Any}, ::Type{SparseMatrixCSC}, a, B, Cs...) =
    broadcast_sp(fn, (front(stups)..., (last(stups)..., a)), mats, containertype(B), containertype(Cs...), B, Cs...)
broadcast_sp(fn, stups, mats, ::Type{SparseMatrixCSC}, ::Type{SparseMatrixCSC}, A, B, Cs...) =
    broadcast_sp(fn, (stups..., ()), (mats..., A), containertype(B), containertype(Cs...), B, Cs...)
# Base cases for broadcast_sp
broadcast_sp(fn, stups, mats, ::Type{SparseMatrixCSC}, ::Type{SparseMatrixCSC}, A, B) =
    broadcast(closeoverscalars((x...) -> fn(front(front(x))..., last(stups)..., last(front(x)), last(x)), front(stups)), mats..., A, B)
broadcast_sp(fn, stups, mats, ::Type{SparseMatrixCSC}, ::Type{Any}, A, bs...) =
    broadcast(closeoverscalars((x...) -> fn(front(x)..., last(stups)..., last(x), bs...), front(stups)), mats..., A)
broadcast_sp(fn, stups, mats, ::Type{Any}, ::Type{SparseMatrixCSC}, a, B) =
    broadcast(closeoverscalars((x...) -> fn(front(x)..., last(stups)..., a, last(x)), front(stups)), mats..., B)
# Recursively close over scalars
closeoverscalars(fn, stups) = closeoverscalars((x...) -> fn(front(x)..., last(stups)..., last(x)), front(stups))
closeoverscalars(fn, ::Tuple{}) = fn
```

(C) The following version is a simpler realization of the preceding concept, but type inference does not seem to like it (nor a similar version using varags for `groupargs`).
```julia
# Entrypoint
broadcast_c(f, ::Type{SparseMatrixCSC}, args...) =
    (ga = groupargs(args); broadcast(closeoverscalars(f, ga[1]), ga[2]...))
# Entrypoint for groupargs
groupargs(args) = groupargs(((),), (), first(args), tail(args))
# Recursion cases for groupargs
groupargs(stups, mats, nextarg, remargs) =
    groupargs((front(stups)..., (last(stups)..., nextarg)), mats, first(remargs), tail(remargs))
groupargs(stups, mats, nextarg::SparseMatrixCSC, remargs) =
    groupargs((stups..., ()), (mats..., nextarg), first(remargs), tail(remargs))
# Base cases for groupargs
groupargs(stups, mats, lastarg, ::Tuple{}) = ((front(stups)..., (last(stups)..., lastarg)), mats)
groupargs(stups, mats, lastarg::SparseMatrixCSC, ::Tuple{}) = ((stups..., ()), (mats..., lastarg))
# Entrypoint and recursion and base cases for closeoverscalars
closeoverscalars(f, stups) = _closeoverscalars((x...) -> f(x..., last(stups)...), front(stups))
_closeoverscalars(f, stups) = _closeoverscalars((x...) -> f(front(x)..., last(stups)..., last(x)), front(stups))
_closeoverscalars(f, ::Tuple{}) = f
```
Best!

**Comment by Sacha0** [comment]
@stevengj, additional thoughts, or is this in shape to merge? (Separating further work along https://github.com/JuliaLang/julia/pull/19239#pullrequestreview-7339126's line seems the most effective path forward?) Thanks and best!

**Comment by Sacha0** [comment]
Thanks for reviewing / merging!

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: Discussion continued [on discourse](https://discourse.julialang.org/t/towards-broadcast-over-combinations-of-sparse-matrices-and-scalars/910). Best!

**Pull request merged**